### PR TITLE
Lazy load fences

### PR DESCRIPTION
**Pull request by Adjective-Object**
Adds a config manager to lazy load fences on request

This does not functionally affect a full fence check, it's only useful in partial fence checks. This is for scenarios where we only need to check against a subset of fences (e.g. there are import changes with no fence changes, and we only need to check the fences of the importer and imported files)

### CODE REVIEW

File: src/utils/loadConfig.ts [hunk 1]
```diff
@@ -8,23 +8,37 @@ import RawExportRule from '../types/rawConfig/RawExportRule';
 import ConfigSet from '../types/ConfigSet';
 import ExportRule from '../types/config/ExportRule';
 import validateRawConfig from '../validation/validateRawConfig';
+import NormalizedPath from '../types/NormalizedPath';
 
-export default function loadConfig(file: string, configSet: ConfigSet) {
+export function loadConfigFromString(
```
Comment (smikula) [inline src/utils/loadConfig.ts:13]: Looks like this function doesn't need to be exported.

File: src/utils/getConfigsForFile.ts [hunk 2]
```diff
@@ -2,22 +2,24 @@ import * as path from 'path';
...
 import NormalizedPath from '../types/NormalizedPath';
 import normalizePath from './normalizePath';
-import getAllConfigs from './getAllConfigs';
+import getConfigManager from './getConfigManager';
 
-// Returns an array of all the configs that apply to a given file
-export default function getConfigsForFile(filePath: NormalizedPath): Config[] {
-    let allConfigs = getAllConfigs();
-    let configsForFile: Config[] = [];
+export function getConfigPathCandidatesForFile(filePath: NormalizedPath): string[] {
```
Comment (smikula) [inline src/utils/getConfigsForFile.ts:7]: Nit: can you split this out into a separate file?  As coded there's a cycle between this file and `getConfigManager`...which isn't strictly a problem, but is easy to resolve.

File: src/utils/getConfigManager.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            this._getAllConfigs();
+        }
+        return this.configSet;
+    }
+
+    public partialConfigSetForPath(configSourcePath: NormalizedPath): ConfigSet {
+        const partialSet: ConfigSet = {};
+
+        if (this.configSet) {
+            for (let configPathCandidate of getConfigPathCandidatesForFile(configSourcePath)) {
```
Comment (smikula) [inline src/utils/getConfigManager.ts:29]: Nit: you can hoist the call to `getConfigPathCandidatesForFile`

Comment (smikula) [reply src/utils/getConfigManager.ts:29]: Also, can you add some brief comments about what you're doing in these different cases?

File: src/utils/getConfigManager.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    private discoveredPaths: Set<string> = new Set();
+
+    public get all(): ConfigSet {
+        if (this.configSet === null) {
+            this._getAllConfigs();
+        }
+        return this.configSet;
+    }
+
+    public partialConfigSetForPath(configSourcePath: NormalizedPath): ConfigSet {
```
Comment (smikula) [inline src/utils/getConfigManager.ts:25]: `getPartialConfigSetForPath`

**Review by smikula — CHANGES_REQUESTED** [review]
I left some minor comments.

File: src/utils/loadConfig.ts (hunk 1 above)
Comment (Adjective-Object) [reply src/utils/loadConfig.ts:13]: It _will_ get used in [this change](https://github.com/Adjective-Object/good-fences/commit/722050c9b7483fc7854e6af946ba8f37ca859c2f) which depends on #97, and I left it exported here as an oversight. I'll just include it in the change that uses it.

**Comment by Adjective-Object** [comment]
There's no reason .all needs to be a property access as opposed to a getter function so I'm going to update it to getAllConfigs() for consistency.

**Review by smikula — APPROVED** [review]

**Pull request merged**